skimage/util/_flood_fill.py

```diff
@@ -0,0 +1,186 @@
+"""flood_fill.py - inplace flood fill algorithm
+
+This module provides a function to fill all equal (or within tolerance) values
+connected to a given seed point with a different value.
+"""
+
+import numpy as np
+
+
+def flood_fill(image, seed_point, new_value, *, selem=None, connectivity=None,
+               indices=False, tolerance=None, inplace=False):
+    """Perform flood filling on an image.
+
+    Starting at a specific `seed_point`, connected points equal or within
+    `tolerance` of the seed value are found, then set to `new_value`.
+
+    Parameters
+    ----------
+    image : ndarray
+        An n-dimensional array.
+    seed_point : tuple or int
+        The index into `image` to start filling.
+    new_value : `image` type
+        New value to set the entire fill.  This must be chosen in agreement
+        with the dtype of `image`.
+    selem : ndarray, optional
+        A structuring element used to determine the neighborhood of each
+        evaluated pixel. It must contain only 1's and 0's, have the same number
+        of dimensions as `image`. If not given, all adjacent pixels are
+        considered as part of the neighborhood (fully connected).
+    connectivity : int, optional
+        A number used to determine the neighborhood of each evaluated pixel.
+        Adjacent pixels whose squared distance from the center is larger or
+        equal to `connectivity` are considered neighbors. Ignored if
+        `selem` is not None.
+    indices : bool, optional
+        If True, the output will be an array representing indices of local
+        maxima. If False, the output will be an array of 0's and 1's with the
+        same shape as `image`.
+    tolerance : ``selem`` type, optional
+        If None (default), adjacent values must be strictly equal to the
+        initial value of `image` at `seed_point`.  This is fastest.  If a value
+        is given, a comparison will be done at every point and this tolerance
+        on each side of the initial value will also be filled (inclusive).
+    inplace : bool, optional
+        If True, flood filling is applied to `image` inplace and nothing is
+        returned.  If False, the flood filled result is returned without
+        modifying the input `image`.  Ignored if `indices` is True.
```

Comment:
> If True, [...] nothing is returned.

Right now this statement is wrong. If we pass in an empty image the function returns an empty array as well and not None (see line 118).
Why not always return the image? That way we could simplify the control flow in and the behavior of the function and remove the special cases as seen here.

However this is not a strong opinion. I really like the API already, especially that `*` in the function signature! :D

Reply:
Some good points here.  I think the majority of NumPy/SciPy `inplace` functions do not return anything, which is what I was emulating, but I am not against returning the inplace modified image.  The empty array case logic definitely needs to be revised, good catch there.

Reply:
Which functions do you have in mind? I was under the impression that NumPy uses the `out` parameter to provide this kind of functionality and still returns the reference even if `out` is provided, e.g. [mean](https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html). The only place I could find this API in SciPy ([winsorize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mstats.winsorize.html#scipy.stats.mstats.winsorize)) the output is still returned. 

Anyway, your argument is quite good, consistency with NumPy / SciPy is desirable. Right now I'm just not sure which way is the consistent one..

Reply:
@lagru I went looking for the API examples I thought I recalled and can't seem to find any.  There are some instances of this in our hidden helper functions, and Pandas does a decent bit of this sort of thing, but I was on the wrong side of this one.  Very diplomatic response, but you're just right.  Also returning in this case lets users call the function and continue in that line with further NumPy methods.

So thanks for the persistence and I'll be making this change.